logic/src/commonMain/kotlin/com/wire/kalium/logic/data/backend_config/BackendTypeMapper.kt

```diff
@@ -0,0 +1,30 @@
+package com.wire.kalium.logic.data.backend_config
+
+import com.wire.kalium.logic.configuration.BuildType
+import com.wire.kalium.network.tools.BackendConfig
+
+
+class BackendTypeMapper {
+    private val staging: BackendConfig
+        get() = BackendConfig(
+            """"https://staging-nginz-https.zinfra.io"""",
+            """"https://wire-account-staging.zinfra.io"""",
+            """"https://staging-nginz-ssl.zinfra.io"""
+        )
+
+    private val production: BackendConfig
+        get() =
+            BackendConfig(
+                """"https://prod-nginz-https.wire.com"""",
+                """"https://account.wire.com"""",
+                """"https://prod-nginz-ssl.wire.com""""
+            )
+
+    fun toBackendConfig(buildType: BuildType): BackendConfig = when (buildType) {
```

Comment:
My only pet peeve I find here is about this `BuildType` -> `BackendConfig` mapping.

I'm not sure if Kalium should say what backend you should use.

Maybe iOS, Web or even Android will need different logics. It's not Kalium/CoreLogic to decide how our product builds will work.

So I'd say to leave this outside, maybe  add `BackendConfig` static values to the class or make it a Sealed Class or something like that. Example:

```kotlin
data class BackendConfig(
    val apiBaseUrl: String,
    val accountsUrl: String,
    val webSocketUrl: String
){ 
    companion object {
        val PRODUCTION = BackendConfig(
                """"https://prod-nginz-https.wire.com"""",
                """"https://account.wire.com"""",
                """"https://prod-nginz-ssl.wire.com""""
            )
        val STAGING = BackendConfig(
            """"https://staging-nginz-https.zinfra.io"""",
            """"https://wire-account-staging.zinfra.io"""",
            """"https://staging-nginz-ssl.zinfra.io"""
        )
    }
}

``` 

And then, when creating the `CoreLogic` whoever is using can just pass `BackendConfig("my", "custom", "values")` or `BackendConfig.PRODUCTION`, for example..

Reply:
+1 for this.

For federation for example, we added the option to switch between testing federation backends in the iOS client.

Reply:
the `BuildType` is suplied by the cosumer app of `kalium` and it's not part of kalium logic e.g.
```kotlin
class CoreLogicCommon(
    protected val buildType: BuildType
) {
```
and the android app will pass this bit of information.
I wanted to make kalium aware of the `BuildType` because it will not only be used for the BEConfig but also can be used for logging (if we want add it in the future)
so my idea was the `BuildType` will ditermin which BE to use, similer to the old implementation in reloaded
```kotlin
object NetworkConfig {
    val properties = mapOf(

        //Config field values for DEBUG Build Type
        BuildTypes.DEBUG to mapOf(
            ConfigFields.API_BASE_URL to ConfigFields.API_BASE_URL.defaultValue,
            ConfigFields.ACCOUNTS_URL to ConfigFields.ACCOUNTS_URL.defaultValue,
            ConfigFields.WEB_SOCKET_URL to ConfigFields.WEB_SOCKET_URL.defaultValue
        ),
        //Config field values for RELEASE Build Type
        //TODO: Certificate pinning, change backend based on flavour
        BuildTypes.RELEASE to mapOf(
            ConfigFields.API_BASE_URL to """"https://prod-nginz-https.wire.com"""",
            ConfigFields.ACCOUNTS_URL to """"https://account.wire.com"""",
            ConfigFields.WEB_SOCKET_URL to """"https://prod-nginz-ssl.wire.com/await?client=""""
        )
    )
}
```

Reply:
OK, I see your point but I think this this is a but inflexible for example in the  use case I mentioned above we added support for running the regular iOS DEV build against a federation backend. In the system you propose the iOS app would need to supply `Build.Custom` which means it loses the ability collect debug logs.

Would it make sense to initialise `CoreLogic` with a `BuildType` and an optional `BackendConfig` which if omitted can default to staging or release depending on the BuildType. This way the client is free to run against production with logging enabled.

Reply:
@typfel I had a discussion with @MohamadJaara and @mchenani about this.

We agreed to leave the concept of `BuildType` out of Kalium and leave that to the apps to figure out.

Also, we had an idea on how to tackle custom backends (and switching backends using deeplinks).

Not 100% correct, but the rough idea is:
```kotlin
// Initialize CoreLogic with ClientInfo and whatever else is needed by the platform.
// ClientInfo will have backend information, URLs, etc.

val core = CoreLogic(someClientInfo)

// All operations here will use the default backend, that was setup before
val sessionAtDefaultBackend = core.authenticationScope {
    loginUsingEmail(email, password).session
}

// The session contains tokens AND information about the backend (urls, etc)


// When creating the userScope, CoreLogic will create HttpClients based on the session info
core.userScope(sessionAtDefaultBackend) {
    conversation.sendTetMessage(params)
}


// User clicks on a deeplink to login into a custom backend

val sessionAtCustomBackend = core.authenticationScope(backendOverwriteData = backendData) {
    // All operations here will use this backend overwrite
    loginUsingEmail(email, password).session
}

core.userScope(sessionAtCustomBackend) {
    // This will use the custom backend, without the need to do anything else.
    conversation.sendTetMessage(params)
}
```

Reply:
@MohamadJaara @typfel thinking about it, what if we just get rid of the default backend info and make the `backendData` mandatory for when getting a `authenticationScope`?

This would make the CoreLogic API easier to understand when dealing with any backend, maybe?

Reply:
Already did it in a similar way, so when invoking `LoginUseCase` `ServerConfig` must be provided regardless if it's on premise or wire BE both are treated the same by `CoreLogic` and after a successful login the user session + `ServerConfig` (the one provided when logging in) will be stored and used by `UserSessionScope` (`CoreLogic` will fetch it internaly from user prefrences).
we can also store the name on the `ServerConfig` since the json data will provide a title as well (in case there is a need to know the BE type).
This approche will allow us to support multible acounts connected to diffrent BE out of the box